Monster nightvision, move vision from flag to json

Coolthulhu opened the pull request:
Monster nightvision adds a lower bound on monster's sight range.

Moved monster vision from VISx flags to vision_dec json. vision_dec kinda works in reverse, but that makes it a bit more intuitive when it comes to darkness - VISx flags are actually more like VIS_PENALTY(60-x) when the light isn't full daylight (60).
Old flags are still handled, just in a different place (at monster generation). Mostly for mod compatibility, but I also didn't update entire monsters.json yet.

Added the nightvision to 3 zeds to show how it works (and to flesh them out a bit): 
- To decayed zombie (40 range, equal to its vision), because it is basically a renamed weaker zombie with no unique role in game
- To rot-weiler (20 range) because it is the dog version of decayed zombie
- To shrieker (10 range), because it fits its role amazingly well.

Also added a mod that gives all zombies nightvision equal to their vision range. Many zombie depictions in media include the ability to locate people in total darkness.

Didn't add nightvision to bots (which, I've been told, are supposed to use IR vision), because they have no good "safety protocols" to prevent gargantuan amounts of player rage that would happen if bots were sniping people from the darkness.

Snaaty commented:
Do the values refer to squares?

So shrieker zombies are now able to see me from ten squares away? If so, I kinda oppose that change, as it makes night-raiding much more difficult than it already is.

Coolthulhu commented:
I can remove the entries from monster.json and just leave it all in the mod if it's too hard.

10 tiles isn't much, though.
And night raiding is incredibly easy. There's next to zero tension, because you can hear everything before it can perceive you in any way. With shriekers having some NV, night raids wouldn't be an instant-win, just a big boost to safety.

Coolthulhu commented:
OK, removed NV from the zeds for now. Back to trivial night raids (at least without the mod).

Changed all the VISx flags to use the field.

Added NV to cats.

RadaRadaRada commented:
Don't forget the big cats: cougar and zombie cougar. Additionally, night vision would be very appropriate for feral hunter zombie.

kevingranade commented:
I'm for monster nv, but cap it to just short of maximum player nv.
Achieving full nv should still be rewarding, but players should fear the
night.

Rivet-the-Zombie commented:
*Most* wild critters could do with some degree of 'night vision' since most of them can see in the dark very, very well; all members of family Cervidae in particular.

Zireael07 commented:
Seconding Kevin's and Rivet's comments, can't wait for it to make it in!

Coolthulhu commented:
Feral hunter and cougar types make sense thematically, but those are already hard to escape from.

I see NVs like this:
Hulks, brutes, shockers, spitters and hunters none - those are too fast and/or ranged.
Dogs low, rot-weilers medium - they're fast
Shriekers, bloated and boomers mid-low - letting them creep up on the player would make them more useful
Stronger regular zeds (cops, soldiers etc.) mid-low
Regular zeds medium
Decayed, crawling - a lot, maybe even more than full player NV to give them a role in game
Bio-ops - none or a lot - it would make sense thematically for them to see in the dark, they shine so it would be obvious when they are attacking, but they are also arguably the strongest zombie type

Rivet-the-Zombie commented:
Our zombies rely on strength of numbers more than anything, and their senses are generally quite weak (except for sense of smell) so I'm not sure why they (with the exception of maybe the bio-op ones) should have better night vision than a living person, other than for game-y reasons.

Coolthulhu commented:
Mostly game-y reasons here, yes.
Night raids are currently rather safe, it just doesn't "feel right" to me that a survivor just strolls through the utter darkness like there's (almost) nothing out there.
Even stumbling on a shrieker isn't dangerous, because you can just run in the direction with the least footstep sounds and soon you're totally safe.

I wanted to strike a balance between the current situation and something that would lead to player rage.

kevingranade commented:
I'm not a fan of giving specific buffs based on balance issues.  There are
plenty of thematic things to do with various possibly underpowered monsters
without resorting to giving them nv just because they're weak.

Dog/cat zombies could get it because they inherited from the host.
Shriekers would specialize for it, same with hunters.
Bio-ops would likely have a nv CBM.

I'm intrigued by the idea of light-emitting zombies.

Coolthulhu commented:
Rotting zombies could be re-themed as specialized night zombies. Pale flesh damaged by sunlight rather than just regular rot.
It's less about just buffing them because they're weak and more about giving them a role, because both gameplay-wise and thematically they're barely different from the much more common plain zeds.

Different NV for different zeds or lump them all in "no NV" and "just below max player NV"?
Shriekers and hunters are human-based, so I think it would be expected that they'd have less (half?) the animal zed NV.

As for light emitting zombies: maybe luminescent boomer bile?
Boomers kinda become obsolete once the player acquires one of the (common) glasses. This would both make them dangerous later on and imitate the behavior that l4d boomers cause in zombies (though through different senses).
Similarly to l4d, it could also be only active when covering living flesh - this would give players incentive to get environmental protection on torso and arms.

BevapDin commented on data/mods/Zombie_Nightvision/main.lua:
```diff
@@ -0,0 +1,8 @@
+-- halve the speed of all monsters whose species is "ZOMBIE"
+for _, monster_type in ipairs(monster_types) do
+    if game.monster_type(monster_type):in_species("ZOMBIE") then
+        game.monster_type(monster_type).vision_min = 60 - game.monster_type(monster_type).vision_dec
```
For this to work, you need to export those properties from `mtype` to Lua. As it's only a public class member, you need to add something like this (and similar for `vision_min`) to the mtype part of "lua/class_definitions.lua" and than compile the game with `make LUA=1` and test the changes:
```
            vision_dec = {
                type = "int",
                writable = true
            },
```

BevapDin commented on src/monstergenerator.cpp:
```diff
@@ -431,6 +433,19 @@ void MonsterGenerator::load_monster(JsonObject &jo)
         newmon->fear = get_set_from_tags(fear_trig, trigger_map, MTRIG_NULL);
         newmon->placate = get_set_from_tags(placate_trig, trigger_map, MTRIG_NULL);
 
+        // Handle legacy flags
```
It would probably better to completely remove those flags. Also remove the from the documentation in doc/JSON_FLAGS.md and please add documentation of the new json members to doc/JSON_INFO.md

Btw. vision_min is not really night vision, as it applies during day as well.

kevingranade commented:
Probably some variation in zombie night vision levels, but the only thing
I'm insistent on is that the best monster nv be a step worse than the best
player nv.
I really like luminescent boomer bile, it's a very reasonable way for
boomers to be really dangerous in certain circumstances.  Basically it'd
try to infect some number of body parts, and inflict a condition that
causes the player to glow.

KA101 commented:
Most zeds shouldn't have it at all, there's no reason for 'em to.  The blob (piloting zeds) being able to see at all is worthwhile on their front.  Felines & zombie felines, sure, and the same for other critters that can see at night.  Shriekers, perhaps.

Bioluminescence is Funky and I'd need a good description of how it works, inasmuch as we'd need to explain it about a few hundred times when people ask on the forums.

Bionic operators need not have NV at all, and if they did it'd be fairly short as their shocker-glow would throw off the compensators anyway.

Decayed zeds are fine as is.  Sometimes it's nice to have a critter you can curbstomp even when starting out.  (Remember, there are players in steps 1-3 of John-Candlebury's cartoon.  They play the game too.)

*You can't assume the player will have, or want to mutate, night vision upgrades*.  Mutation is an optional part of the game.  Any change based on presuming the player will have FNV, and therefore make night-operating v. zeds trivial, is a dealbreaker.

Lain- commented:
I'd like to add that not all zombie must have a "role", some are just there due to thematic reasons (the decayed and crawling zombie) to show that not all your run-of-the-mill zombie are 100% identical.
I'm not against adding some uniquesness to them though, decayed's bites can have higher chance to infect, crawling can "sneak" near player because crawling + terrain. Off-topic here but I want to pump this when talking about monster's ability: Issue #9497

Back to topic, night vision for zom-animal and feral hunter makes sense, though not sure everyone will happy with their night raid being "rudely" interrupted. Some sort of a luminescent-inflicting zombie is nice too, I don't like boomer handle that ability though, if anything I'd vote for a buff to it's existing cabability. Why not make a entirely new zombie and slap that ability on it? AFAIK there are no policy against adding new monster right? Proptosis (bulging-eye) zombie anyone?

Coolthulhu commented:
OK, I added NV for cougars, zougars, shriekers and hunters.
10 tiles, as compared to full nightvision's 12.

If someone wants to add more later on (for example, for deer mentioned above), balanced things etc. it's a trivial json change.

i2amroy commented:
To be honest I'd rather see "vision" and "night_vision" fields with the game scaling monster vision proportionally from "vision" in broad daylight to "night_vision" in absolute darkness automatically. It seems like something like that wouldn't require much extra work, and would allow for much easier access to vision ranges for modders.

(As well as allowing for some new possibilities, like nighttime hunters that can see far and will hunt you during the night, but are extremely blind during the day).

Coolthulhu commented:
OK, changed to vision_day and vision_night.

During the dawn and the evening, sight range is linearly interpolated between those.

KA101 commented:
>OK, I added NV for cougars, zougars, shriekers and hunters.
10 tiles, as compared to full nightvision's 12.

Hmm.

>Any change based on presuming the player will have FNV, and therefore make night-operating v. zeds trivial, is a dealbreaker.

Coolthulhu commented:
So which one then?
>I'm for monster nv, but cap it to just short of maximum player nv. Achieving full nv should still be rewarding, but players should fear the night.

Or

>Any change based on presuming the player will have FNV, and therefore make night-operating v. zeds trivial, is a dealbreaker.

?

Unless you mean that it makes FNV too good, but FNV is much harder to achieve than being able to reliably wreck 3 hulks at once.

KA101 commented:
I'm thinking that demanding FNV is overkill.  Kevin can overrule me if he likes, but I'm hard-pressed to justify giving zeds NV ranged based on a player spec that most mutation lines don't get.

Coolthulhu commented:
I don't think it's anywhere near "demanding". Having some zombies be able to see you before you see them promotes thinking, knocking on the doors/walls (one hint Zombie Survival Guide got right), using noise emitting items to move the horde before going in etc.

Plus, it's just 3 kinds of zed, not one of which is a big player killer.

KA101 commented:
*snort*

Current daylight balance is such that no critter whatsoever can see farther than the player.  Folks flip out about simultaneous sight already, and that's a function of movement.  I'd imagine giving NV would *make* them player killers.

You want felines to have night vision, I don't mind giving them feline-NV.  Full NV is for folks who can expect to spend a significant amount of time underground.

Coolthulhu commented:
Stumbling upon a zed in the middle of the night is expected. Some NV would make certain kinds be stumbled upon more. If not 10, then how much do you think would be good?

I don't think balancing the entire game around the worst players (who don't even night raid and that's why they die) is a good idea. The most common complaint I keep seeing about DDA is its difficulty (lack of thereof).

Snaaty commented:
I appreciate those changes, Coolthulu.

Perhaps those changes could be packed into a mod for now? I can't test it because I can't compile, but I find it rather hard to estimate the increase of difficulty that would take place with these changes. I imagine a shrieker zombie chasing you down while repeatedly shrieking could really mess things up for a number of players. On the other hand, I am not sure what the range of the shriek-ability is. 

On the other hand, personally I'd be fine with merging this and just see what happens, as the values are jsonized it's really easy to tweak them in case someone isn't happy with it.

Coolthulhu commented:
OK, cut out zed NV for now.
Left cougars with NV, because otherwise it wouldn't really be testable.

KA101 commented:
Feline NV is four tiles; 5 if you somehow have Avian Eyes as well.  That's the same as High NV or Ursine Vision.  Elf-A NV is 6 tiles and FNV/Fey Nightsight/Cephalopod Vision/RM13 light-amp/NV-goggles is 12. +1 if you have Avian Eyes.

As I said, I've no problem with feline/ursine mutants having the same night vision as their counterparts, and that goes for all critter-types with NV.  My problem is pegging critter NV to "just a little worse than the best NV the player can get", as that's an arbitrarily game-based standard.

Hunter/shrieker zeds might have some NV, but even one or two tiles would make them a much greater threat than they currently are, so I'm thinking that needs to show up on the forums first.

Rivet-the-Zombie commented:
>/shrieker zeds might have some NV, but even one or two tiles would make them a much greater threat than they currently are

What if shriekers would scream if they got a good sniff of the player's scent? It makes a degree of sense (Hey guys, I smell a live one over here!), serves the same purpose, and would avoid the whole issue of 'why does a zombie specializing in screaming also have magic eyeballs?' at the same time.

Coolthulhu commented:
OK, I'll make a forum thread before sending a balancing PR.
For now I left this one as "feature PR" rather than feature+balancing.

Left cats and cougars with NV - cats at 20 tiles, cougars at 10. Otherwise it wouldn't be testable, only with the zombie NV mod (which has a bit too huge ranges to look good).
Chosen arbitrarily, though I don't think it should be tied to player NV - player doesn't become a cat, just a mutant with cat-like features.

I think it's ready for testing. Passed tests on my side, at least.
If you disagree with cat NV, delete (or adjust) lines with "vision_night" from monsters.json.

KA101 commented:
Was testing cougars in the dark: had one segfault when moving around outside my vision, attached a debugger, and got http://pastebin.com/WA5Tn1Tb when moving through a lab trying to replicate.

Not precisely what I'd anticipated (thinking the jumping when out-of-vision was the problem?) but was weird.

Coolthulhu commented:
I scanned the relevant files trying to find how it could happen and the only explanation that I could find is as follows:
Moving critter creates footsteps (adds itself to footstep_source).
game.cpp invokes cleanup_dead() after all critters move, but before drawing footsteps
If one critter moves and other then kills it (it has to be a creature or possibly NPC doing the killing), the footstep_source won't be a valid pointer.

If the above is true, then monster infighting exacerbated the problem, but didn't cause it.
I managed to get some strange bugs related to monsters fighting at the edge of the bubble, such as hulks swinging at <npcname> and a segfault.

I'm pretty sure it's not related to vision and nightvision, but rather footsteps, reality bubble and monsters unloading/dying.

KA101 commented:
Yeah, that's what I'm thinking.  Been trying to reproduce the original circumstances and no luck on this end.
